bzzeth/bzzeth.go

```diff
@@ -326,6 +349,249 @@ func newChunk(data []byte) chunk.Chunk {
+	defer cancel()
+
+	// missingHeaders collects hashes of headers not found within swarm
+	// ie., the hashes to request from the eth full nodes
+	missingHeaders := make(chan []byte)
+	defer close(missingHeaders)
+	var wg sync.WaitGroup
+
+	// fall back to retrieval from eth clients
+	// collect missing block header hashes
+	// terminates after missingHeaders is read and closed or context is done
+	wg.Add(1)
+	go b.getBlockHeadersEth(ctx, missingHeaders, deliveries, &wg, rcvdPeer)
+
+BZZ:
+	for _, h := range hashes {
+		wg.Add(1)
+		go func(hdr chunk.Address) {
+			defer wg.Done()
+			header, err := b.getBlockHeaderBzz(ctx, hdr)
+			if err != nil {
+				log.Debug("bzzeth.requestAll: netstore.Get can not retrieve chunk", "ref", hex.EncodeToString(h), "err", err)
+				select {
+				case missingHeaders <- hdr: // fallback: request header from eth peers
+				case <-ctx.Done():
+				}
+				return
+			}
+			// deliver the headers received from Swarm
+			select {
+			case deliveries <- header:
+			case <-ctx.Done():
+			}
+		}(h)
+
+		select {
+		case <-ctx.Done():
+			break BZZ
+		default:
+		}
+	}
+
+	// wait till all hashes are requested from swarm OR from another eth node,
+	wg.Wait()
+
+}
+
+// getBlockHeadersEth manages fetching headers from ethereum bzzeth nodes
+// This is part of the response to GetBlockHeaders requests by bzzeth light/syncing nodes
+// As a fallback after header retrieval from local storage and swarm network are unsuccessful
+// When called, it
+// - reads requested header hashes from a channel (headerC) and
+// - creates batch requests and sends them to an adequate bzzeth peer
+// - channels the responses into a delivery channel (deliveries)
+func (b *BzzEth) getBlockHeadersEth(ctx context.Context, headersC, giveBackC chan []byte, pwg *sync.WaitGroup, rcvdPeer *Peer) {
+	log.Debug("getting missing headers from another ETH node", "count", len(giveBackC))
+	defer pwg.Done() // unblock the parent so that i can continue
+
+	// read header requests into batches
+	readNext := make(chan chan []chunk.Address)
+	batches := make(chan []chunk.Address)
+	go readToBatches(headersC, readNext)
+	readNext <- batches
+
+	// send GetBlockHeader requests to adequate bzzeth peers
+	// this loop terminates when batches channel is closed as a result of input headersC being closed
+	var wg sync.WaitGroup
+	deliveryC := make(chan []byte)
+	defer close(deliveryC)
+	for headers := range batches {
+		p := b.peers.getEth(rcvdPeer) // find candidate peer to serve the headers
+		if p == nil {                 // if no peer found just skip the batch TODO: smarter retry?
+			continue
+		}
+		// initiate request with the chosen peer
+		req, err := p.getBlockHeaders(ctx, headers, deliveryC, giveBackC)
+		if err != nil { // in case of failure, no retries TODO: smarter retry?
+			continue
+		}
+
+		wg.Add(1)
+		go b.getDelivery(ctx, req, &wg)
+	}
+	wg.Wait()
+
+	fmt.Println("Finishing getBlockHeadersEth")
```

Comment:
A debug line that should be removed, I suppose.

Reply:
oops.. removed